plans/phase-1/api-design/multigres-operator-api-v1alpha1-design.md

```diff
@@ -0,0 +1,1218 @@
+
+#### `CoreTemplate`, `CellTemplate`, `ShardTemplate`
+
+  * **In-Use Protection (Finalizer):**
+    1.  These templates are protected by a "fan-out" finalizer pattern to prevent accidental deletion of templates actively used by production clusters.
+    2.  **Logic:** When a `MultigresCluster` controller reconciles and sees it is using "prod-cell-template", it adds a finalizer (e.g., `multigres.com/in-use-by-example-cluster`) *to the `CellTemplate` object*.
+    3.  **Cleanup:** When the `MultigresCluster` is deleted (or updated to stop using that template), its controller removes that specific finalizer from the template.
+    4.  **Result:** Kubernetes will reject the deletion of a Template if any Cluster is still referencing it (because the finalizers won't be empty).
+
+## End-User Examples
+
+### 1\. The Ultra-Minimalist (Relying on Namespace/Webhook Defaults)
+
+This creates Multigres cluster with one cell and one database, one tablegroup and one shard.
+
+The defaults for this ultra-minimalistic example can be fetched in two ways:
+
+1.  All components are defaulted by the operator's webhook.
+2.  If a `CoreTemplate`, `CellTemplate`, and `ShardTemplate` named `default` exists within the same namespace it will take these as its default values.
+
+> Notice that the `cells` field is still necessary but we are not naming the cell, this is because we are not sure yet if we should take a default zone or region at random from the cluster to define this, but if we can do this safely this field also won't be needed
+
+```yaml
+apiVersion: multigres.com/v1alpha1
+kind: MultigresCluster
+metadata:
+  name: minimal-cluster
+spec:
+  cells:
+    - zone: "us-east-1a"
+```
+
+When the user does a `kubectl get multigrescluster minimal-cluster -o yaml` after apply this they would see all the values materialized, the default will be applied via webhook.
+
+### 2\. The Minimalist (Relying on Cluster Defaults)
+
+This user relies on the `spec.templateDefaults` field to set cluster-wide defaults.
+
+```yaml
+apiVersion: multigres.com/v1alpha1
+kind: MultigresCluster
+metadata:
+  name: minimal-cluster-with-defaults
+spec:
+  # This cluster will use the "dev-defaults" CoreTemplate for its
+  # global components. All cells and shards will use their respective
+  # "dev-defaults" templates.
+  templateDefaults:
+    coreTemplate: "dev-defaults"
+    cellTemplate: "dev-defaults-cell"
+    shardTemplate: "dev-defaults-shard"
+
+  # CoreComponents (globalTopoServer, multiadmin) are omitted,
+  # so they will use "dev-defaults" (from the CoreTemplate)
+
+  cells:
+    - name: "us-east-1a"
+      zone: "us-east-1a"
+      # 'spec' and 'cellTemplate' are omitted, so "dev-defaults-cell" is used.
+    - name: "us-west-2a"
+      zone: "us-west-2a"
+      # 'spec' and 'cellTemplate' are omitted, so "dev-defaults-cell" is used.
+
+  databases:
+    - name: "db1"
+      tablegroups:
+      - name: "tg1"
+        shards:
```

Comment:
Where is the Multigres's default database + tablegroup? Are they going to be added by mutating webhook?

Reply:
I'm not sure what you mean by the hybrid approach. Do you mean that when it is not specified, the mutating webhook will be placing the default database + tablegroup to the spec anyways? If that's what you mean, I like that approach -- I just think it's better to have explicit update to the spec, rather than controller handling the update behind the scenes.

Reply:
Yes, that's what I mean.